### PR TITLE
[video_player] Added fullscreen functionality (for web).

### PR DESCRIPTION
**Pull request by rsegecin**
Adds a feature on the video_player package to enable a video to go fullscreen on web platform.

https://github.com/flutter/flutter/issues/64397

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart [hunk 1]
```diff
@@ -73,6 +73,16 @@ abstract class VideoPlayerPlatform extends PlatformInterface {
     throw UnimplementedError('pause() has not been implemented.');
   }
 
+  /// The video enters fullScreen (only available for web platform).
```
Comment (stuartmorgan) [inline packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:76]: Functionality that is only ever expected to be relevant for a single platform should be added directly to the platform in question, not the platform interface.

It's not clear to me how this is intended to be used though, as `textureId` is a hidden property of the controller in the app-facing package. What's the user-facing design for this feature?

**Review by stuartmorgan — COMMENTED** [review]
Thanks for the contribution! This will need tests, per the comment above, before moving forward. See also my comment below about the feature design, as it doesn't seem like this is a complete feature as presented.

File: packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart (hunk 1 above)
Comment (rsegecin) [reply packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:76]: Hi Stuart, I've added the rest of the code, fullscreen implemented in the facing app and platform tests. Please review, and let me know about the proceeding steps.

File: packages/video_player/video_player/example/integration_test/video_player_test.dart [hunk 2]
```diff
@@ -91,6 +91,32 @@ void main() {
...
+        // Mute to allow playing without DOM interaction on Web.
+        // See https://developers.google.com/web/updates/2017/09/autoplay-policy-changes
+        await controller.setVolume(0);
+
+        await controller.play();
+        await controller.toggleFullScreen();
+        await tester.pumpAndSettle(_playDuration);
+
+        expect(controller.value.isFullScreen, true);
+        expect(controller.value.position,
```
Comment (stuartmorgan) [inline packages/video_player/video_player/example/integration_test/video_player_test.dart:107]: How is this expectation related to the functionality you are testing?

File: packages/video_player/video_player/example/integration_test/video_player_test.dart [hunk 3]
```diff
@@ -91,6 +91,32 @@ void main() {
...
+            (Duration position) => position > Duration.zero);
+
+        await controller.toggleFullScreen();
+        await tester.pumpAndSettle(_playDuration);
+
+        expect(controller.value.isFullScreen, false);
+        expect(controller.value.position,
```
Comment (stuartmorgan) [inline packages/video_player/video_player/example/integration_test/video_player_test.dart:114]: Same question here.

File: packages/video_player/video_player/example/integration_test/video_player_test.dart [hunk 4]
```diff
@@ -91,6 +91,32 @@ void main() {
...
+      'can go fullscreen',
+      (WidgetTester tester) async {
+        await controller.initialize();
+        // Mute to allow playing without DOM interaction on Web.
+        // See https://developers.google.com/web/updates/2017/09/autoplay-policy-changes
+        await controller.setVolume(0);
+
+        await controller.play();
+        await controller.toggleFullScreen();
+        await tester.pumpAndSettle(_playDuration);
```
Comment (stuartmorgan) [inline packages/video_player/video_player/example/integration_test/video_player_test.dart:104]: Why do you have to play a second of video for full screen changes to take effect?

File: packages/video_player/video_player/example/integration_test/video_player_test.dart [hunk 5]
```diff
@@ -91,6 +91,32 @@ void main() {
...
+            (Duration position) => position > Duration.zero);
+
+        await controller.toggleFullScreen();
+        await tester.pumpAndSettle(_playDuration);
+
+        expect(controller.value.isFullScreen, false);
+        expect(controller.value.position,
+            (Duration position) => position > Duration.zero);
+      },
+      skip: !kIsWeb,
```
Comment (stuartmorgan) [inline packages/video_player/video_player/example/integration_test/video_player_test.dart:117]: Rather than making this platform-gated, there should be a `supportsFullScreen` API at both the app-facing and platform interface levels so that it's runtime-checkable.

File: packages/video_player/video_player/lib/video_player.dart [hunk 6]
```diff
@@ -475,6 +481,19 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
     await _applyPlayPause();
   }
 
+  /// The video enters in fullscreen (only works for Web).
+  Future<void> toggleFullScreen() async {
+    if (kIsWeb) {
+      if (!value.isFullScreen) {
+        await _applyFullScreen();
```
Comment (stuartmorgan) [inline packages/video_player/video_player/lib/video_player.dart:488]: Since these are only used in one place, they don't need to be separated out into helpers.

File: packages/video_player/video_player/lib/video_player.dart [hunk 7]
```diff
@@ -475,6 +481,19 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
...
+        value = value.copyWith(isFullScreen: true);
+      } else {
+        await _applyExitFullScreen();
+        value = value.copyWith(isFullScreen: false);
```
Comment (stuartmorgan) [inline packages/video_player/video_player/lib/video_player.dart:492]: Can't the user exit full screen directly? It seems like we would need a state listener for this, rather than managing the state from Dart.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/video_player/video_player_web/example/integration_test/video_player_web_test.dart [hunk 8]
```diff
@@ -125,6 +125,16 @@ void main() {
       expect(VideoPlayerPlatform.instance.pause(await textureId), completes);
     });
 
+    testWidgets('can enter fullscreen', (WidgetTester tester) async {
+      expect(VideoPlayerPlatform.instance.enterFullScreen(await textureId),
+          completes);
```
Comment (stuartmorgan) [inline packages/video_player/video_player_web/example/integration_test/video_player_web_test.dart:130]: This is only testing that the future completes, not that it does anything. Wouldn't at empty implementation of `enterFullScreen` pass this test?

**Comment by stuartmorgan** [comment]
@rsegecin Are you still planning on updating this PR based on the feedback above?

**Comment by rsegecin** [comment]
> @rsegecin Are you still planning on updating this PR based on the feedback above?

Yes, I do want to continue but I think it'll take some time for me to understand, learn and make the appropriate corrections you mentioned. I've been busy with other projects and I don't know when I'll work in this PR so if you rather close it I'll understand may be we can reopen in the future right? 

Btw thank you very much for your review @stuartmorgan.

**Comment by stuartmorgan** [comment]
I'll mark it as a draft for now so that it doesn't show up in our regular PR triage; please feel free to mark it as ready for review again once you've updated it.

**Comment by ditman** [comment]
This PR is adding support to enable the native full-screen support on web, without needing to add a controller-owned method:

* https://github.com/flutter/packages/pull/3259

(The difference that I see is that apps cannot know/trigger the full-screen mode, it needs to be user-triggered from the native UI)

@rsegecin what do you think about 3259, would it fulfill your needs?

**Comment by rsegecin** [comment]
> This PR is adding support to enable the native full-screen support on web, without needing to add a controller-owned method:
> 
> * [[video_player] Add optional web options #3259](https://github.com/flutter/packages/pull/3259)
> 
> (The difference that I see is that apps cannot know/trigger the full-screen mode, it needs to be user-triggered from the native UI)
> 
> @rsegecin what do you think about 3259, would it fulfill your needs?

Yes, definitely, thank you for notifying me @ditman. Kudos to @defuncart coming up with this nice way to set web options for video_player.
